pkg/controllers/applicationfailover/crb_application_failover_controller.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/karmada-io/karmada/pkg/features"
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+// CRBApplicationFailoverControllerName is the controller name that will be used when reporting events.
+const CRBApplicationFailoverControllerName = "cluster-resource-binding-application-failover-controller"
+
+// CRBApplicationFailoverController is to sync ClusterResourceBinding's application failover behavior.
+type CRBApplicationFailoverController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+
+	// workloadUnhealthyMap records which clusters the specific resource is in an unhealthy state
+	workloadUnhealthyMap *workloadUnhealthyMap
+	ResourceInterpreter  resourceinterpreter.ResourceInterpreter
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CRBApplicationFailoverController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling ClusterResourceBinding %s.", req.Name)
+
+	binding := &workv1alpha2.ClusterResourceBinding{}
+	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !binding.DeletionTimestamp.IsZero() ||
+		binding.Spec.Failover == nil || binding.Spec.Failover.Application == nil {
+		resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+		if err != nil {
+			return controllerruntime.Result{Requeue: true}, err
+		}
+		c.workloadUnhealthyMap.delete(resource)
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to check health status of the workload after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *CRBApplicationFailoverController) syncBinding(binding *workv1alpha2.ClusterResourceBinding) (time.Duration, error) {
+	// cleanup expired cluster
+	tolerationSeconds := binding.Spec.Failover.Application.DecisionConditions.TolerationSeconds
+	resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+	if err != nil {
+		klog.Errorf("failed to get key from binding(%s)'s resource", binding.Name)
+		return 0, err
+	}
+
+	unhealthyClusters := c.workloadUnhealthyMap.get(resource)
+	if unhealthyClusters == nil {
+		unhealthyClusters = make(map[string]metav1.Time)
+	}
```

Comment:
Why do this initialization in a rush, can we do it after the eviction decision?

Reply:
`UnhealthyClusters` is used to check whether the cluster need to be evicted. I think it need to be inited before detecting failure.